osquery/tables/system/windows/prefetch.cpp

```diff
@@ -0,0 +1,307 @@
+    timestamps.push_back(time);
+    run_times.erase(run_times.begin());
+    times++;
+  }
+  std::string timestamp_list = osquery::join(timestamps, ",");
+  int run_count = 0;
+  if (version == 23) {
+    run_count = *reinterpret_cast<int*>(&prefetch_data[152]);
+  } else if (version == 26) {
+    run_count = *reinterpret_cast<int*>(&prefetch_data[208]);
+  } else {
+    run_count = *reinterpret_cast<int*>(&prefetch_data[200]);
+  }
+
+  auto r = make_table_row();
+  r["path"] = file_path;
+  r["filename"] = header.filename;
+  r["hash"] = header.prefetch_hash;
+  r["size"] = INTEGER(header.file_size);
+  r["number_of_accessed_files"] = INTEGER(accessed_file_list.size());
+  r["accessed_files"] = files_accessed_list;
+  r["volume_serial"] = osquery::join(volume_serial_list, ",");
+  r["volume_creation"] = osquery::join(volume_creation_list, ",");
+  r["number_of_accessed_directories"] = INTEGER(accessed_dirs_list.size());
+  r["accessed_directories"] = dirs_accessed_list;
+  r["last_execution_time"] = INTEGER(timestamps[0]);
+  r["execution_times"] = timestamp_list;
+  r["count"] = INTEGER(run_count);
+  yield(std::move(r));
+}
+
+void parsePrefetchVersion(RowYield& yield,
+                          const std::vector<UCHAR>& data,
+                          const std::string& file_path) {
+  std::vector<UCHAR> prefetch_data = data;
+  int version = *reinterpret_cast<int*>(&prefetch_data[0]);
+  // Currently supports Win7 and higher
+  if (version == 30 || version == 23 || version == 26) {
+    parsePrefetchData(yield, prefetch_data, file_path, version);
+  } else {
+    LOG(WARNING) << "Unsupported prefetch file: " << file_path;
+  }
+}
+
+void parsePrefetch(const std::vector<std::string>& prefetch_files,
+                   RowYield& yield) {
+  for (const auto& file : prefetch_files) {
+    if (boost::algorithm::iends_with(file, ".pf") &&
+        boost::filesystem::is_regular_file(file)) {
+      LOG(INFO) << "Parsing prefetch file: " << file;
+      std::ifstream input_file(file, std::ios::in | std::ios::binary);
+      std::vector<char> compressed_data(
+          (std::istreambuf_iterator<char>(input_file)),
+          (std::istreambuf_iterator<char>()));
+      input_file.close();
+
+      long header_sig = *reinterpret_cast<long*>(&compressed_data[0]);
+      long size = *reinterpret_cast<long*>(&compressed_data[4]);
+      std::vector<UCHAR> data;
+      std::string file_path = file;
+      // Check for compression signature MAM04. Prefetch may be compressed on
+      // Win8+
+      if (header_sig == 72171853) {
+        auto expected_data = decompressLZxpress(compressed_data, size);
+        if (expected_data.isError()) {
+          continue;
+        }
+        data = expected_data.take();
+      } else {
+        std::transform(
+            compressed_data.begin(),
+            compressed_data.end(),
+            std::back_inserter(data),
+            [](char char_data) { return reinterpret_cast<UCHAR&>(char_data); });
+      }
```

Comment:
What's the average size of a file that is not compressed?
This seems an unnecessary copy, wouldn't it be better to read the data already with UCHAR then change the decompressLZxpress API, so that it accepts a `const std::vector<UCHAR>&` as input?
This way you don't need to do any other copy.

Reply:
good question, im not really sure what the average size of a file is that is not compressed.
Only prefetch files on win7 or earlier are _not_ compressed
Win8+ files are all compressed
But good point about just reading the data with UCHAR. Switched reading file to UCHAR